lib/metasploit/framework/login_scanner/redis.rb

```diff
@@ -58,7 +60,7 @@ def attempt_login(credential)
             # Invalid password - ( -ERR invalid password\r\n )
             # Valid password   - (+OK\r\n)
 
-            if result_options[:proof] && result_options[:proof] =~ /but no password is set/i
```

Comment:
This code block is getting a bit unweildly. Suggest:

```ruby
            if result_options[:proof]
              if result_options[:proof] =~ OLD_PASSWORD_UNSET || result_options[:proof] =~ PASSWORD_UNSET)
                result_options[:status] = Metasploit::Model::Login::Status::NO_AUTH_REQUIRED
              elsif result_options[:proof] =~ /^-ERR invalid password/i
                result_options[:status] = Metasploit::Model::Login::Status::INCORRECT
              elsif result_options[:proof][/^\+OK/]
                result_options[:status] = Metasploit::Model::Login::Status::SUCCESSFUL
              end
            end
```


The `attempt_login` method is also becoming a bit complex. Suggest moving response parsing to another method.


```ruby
        # This method attempts a single login with a single credential against the target
        # @param credential [Credential] The credential object to attempt to login with
        # @return [Metasploit::Framework::LoginScanner::Result] The LoginScanner Result object
        def attempt_login(credential)
          result_options = {
            credential: credential,
            status: Metasploit::Model::Login::Status::INCORRECT,
            host: host,
            port: port,
            protocol: 'tcp',
            service_name: 'redis'
          }

          disconnect if self.sock

          begin
            connect
            select([sock], nil, nil, 0.4)
            sock.put(redis_proto(['AUTH', "#{credential.private}"]))
            proof = validate_login(sock.get_once)
            result_options[:proof] = proof unless proof.nil?
          rescue Rex::ConnectionError, EOFError, Timeout::Error, Errno::EPIPE => e
            result_options.merge!(
              proof: e,
              status: Metasploit::Model::Login::Status::UNABLE_TO_CONNECT
            )
          end

          disconnect if self.sock
          ::Metasploit::Framework::LoginScanner::Result.new(result_options)
        end

        def validate_login(data)
          return if data.nil?

          # No password      - ( -ERR Client sent AUTH, but no password is set\r\n )
          # Invalid password - ( -ERR invalid password\r\n )
          # Valid password   - (+OK\r\n)

          if data =~ OLD_PASSWORD_UNSET
            return Metasploit::Model::Login::Status::NO_AUTH_REQUIRED
          elsif data =~ PASSWORD_UNSET)
            return Metasploit::Model::Login::Status::NO_AUTH_REQUIRED
          elsif result_options[:proof] =~ /^-ERR invalid password/i
            return Metasploit::Model::Login::Status::INCORRECT
          elsif result_options[:proof][/^\+OK/]
            return Metasploit::Model::Login::Status::SUCCESSFUL
          else
            return nil
          end
        end
```

Reply:
cleaned up, thanks for the suggestion.